app/models/qc_report.rb

```diff
@@ -15,9 +15,13 @@ def available_states
     def self.included(base)
       base.class_eval do
         # When adding new states, please make sure you update the config/locals/en.yml file
-        # with decriptions.
+        # with descriptions.
 
-        aasm column: :state, whiny_persistence: true do
+        # We disable the transactions here as otherwise our entire job gets wrapped in a transaction.
+        # This causes problems for very large reports, as it causes objects with after transaction callbacks to
+        # be persisted for the entire job. We'd already chunked our transaction into batches, but this transaction
+        # was counteracting that.
+        aasm column: :state, whiny_persistence: true, use_transactions: false do
```

Comment:
Just to clarify:
Would it do one transaction per state transition, so:
* One for `event :generate`
* One for `state :generating, after_enter: :generate_report`, which stretches on until the `generate_report` function has finished? ( which include the `generation_complete!` event )

i.e. the 'after transaction callbacks' you're referring to is `after_enter: :generate_report` in this case?

Thanks

Reply:
I believe it wraps the entire event and the callbacks on the state transitions as well. There's an outline here:
https://github.com/aasm/aasm#lifecycle
So in this case it would begin at generate, and encompass everything through to the end of generate_report.

I'm not sure if it would open a 'nested' transaction for the 'generation_complete!' event, or if it's happy to use the existing one. I'd need to double check the aasm code.

Reply:
Ahh, actually, I think this behaviour is configurable. You can enforce it opening a new transaction per state transition. https://github.com/aasm/aasm#transaction-support